File: week12/11/2/P2.py
from functools import cmp_to_key

class Solution:
    def largestNumber(self, nums: List[int]) -> str:
        # 사용자 정의 비교 함수
        def compare(x, y):
            if x + y > y + x:
                return -1  # x + y가 더 크면 x가 앞에
            elif x + y < y + x:
                return 1   # y + x가 더 크면 y가 앞에
            else:
                return 0   

        # 숫자를 문자열로 변환하여 정렬
        nums = list(map(str, nums))
        nums.sort(key=cmp_to_key(compare))

        # 결과 결합
        result = "".join(nums)

        # "0"만 포함된 경우 처리
        return "0" if result[0] == "0" else result


    def largestNumber(self, nums:List[int]) -> str:
        nums = list(map(str,nums))

        for i in range(1,len(nums)):
            pivot = nums[i]
            j = i - 1
            # 한칸씩 내리면서 들어가야할 자리를 찾음
            while j >= 0 and nums[j] + pivot < pivot + nums[j]:
                nums[j + 1] = nums[j]
                j -= 1
            nums[j + 1] = pivot
        
        result = "".join(nums)
        return "0" if result[0] == "0" else result

File: week12/11/2/P2.cpp
#include <iostream>
#include <string>

using namespace std;

class Solution {
public:
    string largestNumber(vector<int>& nums) {
        vector<string> renums;
        for (int num : nums) {
            renums.push_back(to_string(num)); // 정수를 문자열로 변환
        }

        for(int i = 1; i< nums.size(); i++){
            string pivot = renums[i];
            int j = i-1;
            while(j >= 0 && renums[j] + pivot < pivot + renums[j]){
                renums[j+1] = renums[j];
                j--;
            }
            renums[j+1] = pivot;
        }

        string result;
        for (string s : renums){
            result += s;
        }
        if (result[0] == '0') return "0";
        return result;
    }

    string largestNumber2(vector<int>& nums) {
        // Step 1: Convert integers to strings
        vector<string> renums;
        for (int num : nums) {
            renums.push_back(to_string(num));
        }

        // Step 2: Sort the strings based on the custom comparator
        sort(renums.begin(), renums.end(), [](const string& a, const string& b) {
            return a + b > b + a; // Compare concatenated results
        });

        // Step 3: Concatenate sorted strings
        string result;
        for (string s : renums) {
            result += s;
        }

        // Step 4: Handle the case where the result is all zeros
        if (result[0] == '0') return "0";

        return result;
    }
};


int main() {
    Solution sol;
    vector<int> nums = {3, 30, 34, 5, 9};
    cout << sol.largestNumber(nums) << endl; // Output: "9534330"
    return 0;
}
